File: slides/salesiano-digital-logic-2023/lista-encadeada/java/Main.java
class Node {
  int data;
  Node next;

  public Node(int data) {
    this.data = data;
    this.next = null;
  }
}

class LinkedList {
  private Node head;

  public LinkedList() {
    this.head = null;
  }

  public void append(int data) {
    Node newNode = new Node(data);
    if (head == null) {
      head = newNode;
    } else {
      Node current = head;
      while (current.next != null) {
        current = current.next;
      }
      current.next = newNode;
    }
  }

  public void remove(int data) {
    if (head != null && head.data == data) {
      head = head.next;
      return;
    }
    Node current = head;
    while (current != null && current.next != null) {
      if (current.next.data == data) {
        current.next = current.next.next;
        return;
      }
      current = current.next;
    }
  }

  public void print() {
    Node current = head;
    while (current != null) {
      System.out.print(current.data + " ");
      current = current.next;
    }
    System.out.println();
  }
}

public class Main {
  public static void main(String[] args) {
    LinkedList linkedList = new LinkedList();
    linkedList.append(1);
    linkedList.append(2);
    linkedList.append(3);

    linkedList.print(); // Output: 1 2 3

    linkedList.remove(2);

    linkedList.print(); // Output: 1 3

  }
}

File: slides/salesiano-digital-logic-2023/lista-encadeada/pascal/main.pas
program LinkedList;

type
  Node = record
    data: Integer;
    next: ^Node;
  end;

type
  LinkedList = record
    head: ^Node;
  end;

procedure InitializeLinkedList(var list: LinkedList);
begin
  list.head := nil;
end;

procedure Append(var list: LinkedList; data: Integer);
var
  newNode, current: ^Node;
begin
  newNode := New(^Node);
  newNode^.data := data;
  newNode^.next := nil;

  if list.head = nil then
    list.head := newNode
  else
  begin
    current := list.head;
    while current^.next <> nil do
      current := current^.next;
    current^.next := newNode;
  end;
end;

procedure Remove(var list: LinkedList; data: Integer);
var
  current, prev: ^Node;
begin
  current := list.head;
  prev := nil;

  while current <> nil do
  begin
    if current^.data = data then
    begin
      if prev = nil then
        list.head := current^.next
      else
        prev^.next := current^.next;
      Dispose(current);
      Exit;
    end;
    prev := current;
    current := current^.next;
  end;
end;

procedure Print(const list: LinkedList);
var
  current: ^Node;
begin
  current := list.head;
  while current <> nil do
  begin
    Write(current^.data, ' ');
    current := current^.next;
  end;
  Writeln;
end;

var
  linkedList: LinkedList;

begin
  InitializeLinkedList(linkedList);

  Append(linkedList, 1);
  Append(linkedList, 2);
  Append(linkedList, 3);

  Print(linkedList);  // Output: 1 2 3

  Remove(linkedList, 2);
  Print(linkedList);  // Output: 1 3
end.

